src/OpenLoco/src/Graphics/SoftwareDrawingContext.cpp

```diff
@@ -726,16 +726,243 @@ namespace OpenLoco::Gfx
             drawImage(*rt, { x, y }, ImageId::fromUInt32(image));
         }
 
+        // 0x00450890, 0x00450F87, 0x00450D1E, 0x00450ABA
+        template<int32_t TZoomLevel>
+        static void drawMaskedZoom(
+            int16_t imageHeight,
+            int16_t imageWidth,
+            int16_t rowSize,
+            const uint8_t* bytesMask,
+            int16_t dstWrap,
+            uint8_t* dstBuf,
+            const uint8_t* bytesImage)
+        {
+            const auto scaledHeight = imageHeight >> TZoomLevel;
+            if (scaledHeight == 0)
+                return;
+
+            const auto scaledWidth = imageWidth >> TZoomLevel;
+            if (scaledWidth == 0)
+                return;
+
+            constexpr auto skip = (1U << TZoomLevel);
+
+            // Calculate the row size.
+            const auto scaledRowSize = [&]() {
+                if constexpr (TZoomLevel == 0)
+                {
+                    return rowSize;
+                }
+                else
+                {
+                    constexpr auto shift = (1 << TZoomLevel) - 1;
+
+                    const auto v7 = shift * static_cast<uint16_t>(rowSize + imageWidth) + rowSize;
+                    const auto v8 = static_cast<uint16_t>(imageWidth) & shift;
+
+                    return v8 + v7;
+                }
+            }();
+
+            if constexpr (TZoomLevel > 0)
+            {
+                dstWrap -= scaledWidth;
+            }
+
+            auto remainingRows = scaledHeight;
+            do
+            {
+                auto remainingWidth = scaledWidth;
+                do
+                {
+                    uint8_t masked = *bytesMask & *bytesImage;
+                    if (masked)
+                        *dstBuf = masked;
+
+                    bytesImage += skip;
+                    bytesMask += skip;
+                    dstBuf++;
+                    remainingWidth--;
+
+                } while (remainingWidth);
+
+                bytesImage += scaledRowSize;
+                bytesMask += scaledRowSize;
+                dstBuf += dstWrap;
+
+                --remainingRows;
+
+            } while (remainingRows);
+        }
+
+        template<int32_t TZoomLevel>
+        static void drawImageMaskedZoom(const RenderTarget& rt, const Ui::Point& pos, const ImageId& image, const ImageId& maskImage)
+        {
+            const auto* g1Image = Gfx::getG1Element(image.getIndex());
+            const auto* g1ImageMask = Gfx::getG1Element(maskImage.getIndex());
```

Comment:
Should probably have some asserts that these images are bmp. Rle Will not work here